atom/browser/ui/tray_icon_cocoa.mm

```diff
@@ -9,81 +9,216 @@
 #include "ui/gfx/image/image.h"
 #include "ui/gfx/screen.h"
 
-@interface StatusItemController : NSObject {
+const CGFloat kStatusItemLength = 26;
+const CGFloat kMargin = 3;
+
+@interface StatusItemView : NSView {
   atom::TrayIconCocoa* trayIcon_; // weak
+  AtomMenuController* menuController_; // weak
+  BOOL isHighlightEnable_;
+  BOOL inMouseEventSequence_;
+  base::scoped_nsobject<NSImage> image_;
+  base::scoped_nsobject<NSImage> alternateImage_;
+  base::scoped_nsobject<NSString> title_;
+  base::scoped_nsobject<NSStatusItem> statusItem_;
 }
-- (id)initWithIcon:(atom::TrayIconCocoa*)icon;
-- (void)handleClick:(id)sender;
-- (void)handleDoubleClick:(id)sender;
 
-@end // @interface StatusItemController
+@end  // @interface StatusItemView
 
-@implementation StatusItemController
+@implementation StatusItemView
 
 - (id)initWithIcon:(atom::TrayIconCocoa*)icon {
   trayIcon_ = icon;
+  isHighlightEnable_ = YES;
+  statusItem_.reset([[[NSStatusBar systemStatusBar] statusItemWithLength:
+      NSVariableStatusItemLength] retain]);
+  NSRect frame = NSMakeRect(0,
+                            0,
+                            kStatusItemLength,
+                            [[statusItem_ statusBar] thickness]);
+  if ((self = [super initWithFrame:frame])) {
+    [statusItem_ setView:self];
+  }
   return self;
 }
 
-- (void)handleClick:(id)sender {
-  // Get the frame of the NSStatusItem.
-  NSRect frame = [NSApp currentEvent].window.frame;
-  gfx::Rect bounds(frame.origin.x, 0, NSWidth(frame), NSHeight(frame));
-  // Flip coordinates to gfx (0,0 in top-left corner) using current screen.
-  NSScreen* screen = [[NSScreen screens] objectAtIndex:0];
-  bounds.set_y(NSHeight([screen frame]) - NSMaxY(frame));
+- (void)removeItem {
+  [[NSStatusBar systemStatusBar] removeStatusItem:statusItem_];
+  statusItem_.reset();
+}
+
+- (void)drawRect:(NSRect)dirtyRect {
+  // Draw the tray icon and title that align with NSSStatusItem, layout:
+  //   ----------------
+  //   | icon | title |
+  ///  ----------------
+  BOOL highlight = [self shouldHighlight];
+  CGFloat titleWidth = [self titleWidth];
+  // Calculate the total icon bounds.
+  NSRect statusItemBounds = NSMakeRect(0,
+                                       0,
+                                       kStatusItemLength + titleWidth,
+                                       [[statusItem_ statusBar] thickness]);
+  [statusItem_ drawStatusBarBackgroundInRect:statusItemBounds
+                               withHighlight:highlight];
```

Comment:
This behavior is consitent with the official. The official behavior will also highlight when alternative image is set and hightlight is enable.